module/zfs/zio.c

```diff
@@ -3252,9 +3252,31 @@ zio_ddt_write(zio_t *zio)
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_table_quota &&
+	    ddt_get_ddt_dsize(spa) >= spa->spa_dedup_table_quota) {
```

Comment:
you might move this logic to ddt.c, with a new function to determine if we are over the ddt quota.

Reply:
Good point. Done :).